app/grandchallenge/cases/models.py

```diff
@@ -996,3 +1056,106 @@ def deidentify_user_uploads(self):
+        try:
+            obj = self._s3_client.get_object(Bucket=bucket, Key=key)
+        except self._s3_client.exceptions.NoSuchKey as e:
+            raise RetryStep("Manifest not (yet) found for job output") from e
+
+        return json.load(obj["Body"])["jobSummary"]
+
+    def get_job_output_failure_log(self, *, job_summary):
+        output_uri = job_summary["failureOutputS3Uri"]
+        parsed = urlparse(output_uri)
+        bucket = parsed.netloc
+        key = parsed.path.lstrip("/") + "failure.ndjson"
+
+        obj = self._s3_client.get_object(Bucket=bucket, Key=key)
+
+        return [json.loads(line) for line in obj["Body"]]
+
+    def handle_event(self, *, event):
+        try:
+            job_status = event["jobStatus"]
+            if job_status == "COMPLETED":
+                self.handle_completed_job(event=event)
+            elif job_status == "FAILED":
+                self.handle_failed_job(event=event)
+            else:
+                raise ValueError("Invalid job status")
+        except Exception as e:
+            self._mark_failed(
+                error_message="An unexpected error occurred", exc=e
+            )
+        else:
+            self.status = self.DICOMImageSetUploadStatusChoices.COMPLETED
+            self.save()
+
+    def handle_completed_job(self, *, event):
+        image_set = self.validate_image_set(event=event)
+        self.convert_image_set_to_internal(image_set=image_set)
+
+    def validate_image_set(self, *, event):
+        job_summary = self.get_job_summary(event=event)
+        if job_summary["numberOfGeneratedImageSets"] == 0:
+            self.handle_failed_job(event=event)
+        if job_summary["numberOfGeneratedImageSets"] > 1:
+            self.delete_image_sets(job_summary=job_summary)
+            raise DICOMImportJobValidationError(
+                "Multiple image sets created. Expected only one."
+            )
+        image_set = job_summary["imageSetsSummary"][0]
+        if not image_set["isPrimary"]:
+            self.delete_image_sets(job_summary=job_summary)
+            raise DICOMImportJobValidationError(
+                "New instance is not primary: "
+                "metadata conflicts with already existing instance."
+            )
+        if not image_set["imageSetVersion"] == 1:
+            self.revert_image_set_to_initial_version(image_set=image_set)
+            raise DICOMImportJobValidationError(
+                "Instance already exists. This should never happen!"
+            )
+        return image_set
+
+    def handle_failed_job(self, *, event):
+        job_summary = self.get_job_summary(event=event)
+        self.internal_failure_log = self.get_job_output_failure_log(
+            job_summary=job_summary
+        )
+        self.save()
+        self.delete_image_sets(job_summary=job_summary)
+        job_id = job_summary["jobId"]
+        raise DICOMImportJobFailedError(message=f"Import job {job_id} failed")
```

Comment:
Maybe add the pk of the DICOMImageSetUpload instance to the error message as well?